File: Server/src/logic/game/Game.java
package logic.game;

import logic.action.Action;
import logic.action.ActionFactory;
import logic.cards.CardData;
import logic.cards.CardManager;
import logic.player.Player;
import logic.player.PlayerAction;

import java.util.List;
import java.util.Stack;

/**
 * Created by Aedan Smith.
 */

public abstract class Game implements Runnable {
    private ActionFactory actionFactory;
    public Stack<Action> stack = new Stack<>();
    public List<Player> players;

    public Game(List<Player> players){
        this.players = players;
        this.actionFactory = new ActionFactory();
    }

    @Override
    public void run() {
        onBegin();
        PlayerIterator turns = new PlayerIterator(players);
        while (!isOver()){
            Player current = turns.next();
            onUpkeep(current);
            resolveStack();
            doTurn(current);
            resolveStack();
            onEndStep(current);
            resolveStack();
        }
    }

    public void resolveStack(){
        while (!stack.isEmpty()){
            stack.pop().apply(this);
            applyStateBasedActions();
        }
        applyStateBasedActions();
    }

    public void applyStateBasedActions(){
        for (Player player : players) {
            player.board.applyStateBasedActions();
        }
    }

    protected abstract void onBegin();

    protected abstract void onUpkeep(Player current);

    protected abstract void onEndStep(Player current);

    private boolean isOver() {
        return false;
    }

    private void doTurn(Player current){
        while (true) {
            resolveStack();
            PlayerAction playerAction = current.nextAction();
            if (playerAction.getActionType() == PlayerAction.ActionType.END_TURN)
                return;
            else if (playerAction.getActionType() == PlayerAction.ActionType.DO_ACTION)
                get(playerAction.getValue()).apply(this, current);
        }
    }

    private GameAction get(String string){
        if (string.charAt(0) == 'h'){
            return new GameAction() {
                @Override
                public void apply(Game game, Player current) {
                    int index = string.charAt(1)-33;
                    CardData cardData = CardManager.get(current.hand.get(index));
                    if (cardData.canCast(current, game)) {
                        current.hand.remove(index);
                        cardData.onCast(current, game);
                    }
                }
            };
        }
        throw new RuntimeException("Could not find action \"" + string + "\"");
    }

    public Object getTarget(String s) {
        switch (s.charAt(0)){
            case 'b':
                return players.get(s.charAt(1)-33).board.getBoard().get(s.charAt(1)-33);
            default:
                throw new RuntimeException("Could not find target \"" + s + "\"");
        }
    }

    public ActionFactory getActionFactory() {
        return actionFactory;
    }
}

File: Server/src/logic/game/Game.kt
package logic.game

import logic.action.Action
import logic.cards.CardManager
import logic.game.gamestates.BeginGameState
import logic.game.gamestates.BeginTurnState
import logic.game.gamestates.DoTurnState
import logic.game.gamestates.EndTurnState
import logic.player.Player
import logic.player.externalplayer.commands
import java.util.*

/**
 * Created by Aedan Smith.
 */

class Game(val players: MutableList<Player>) : Runnable {
    var stack = Stack<Action>()
    var beginGameState = BeginGameState()
    val beginTurnStates = ArrayList<BeginTurnState>(players.size)
    val doTurnStates = ArrayList<DoTurnState>(players.size)
    val endTurnStates = ArrayList<EndTurnState>(players.size)

    init {
        players.forEach { beginTurnStates.add(BeginTurnState(it)) }
        players.forEach { doTurnStates.add(DoTurnState(it)) }
        players.forEach { endTurnStates.add(EndTurnState(it)) }
    }
    
    override fun run() {
        var nextState: GameState = beginGameState
        while (true) {
            nextState = nextState.apply(this)
        }
    }

    fun resolveStack() {
        while (!stack.isEmpty()) {
            stack.pop().apply(this)
            applyStateBasedActions()
        }
        applyStateBasedActions()
    }

    fun applyStateBasedActions() {
        for (player in players) {
            player.board.applyStateBasedActions()
        }
    }

    fun indexOf(player: Player) = players.indexOf(player)

    fun get(string: String): GameAction {
        if (string.startsWith(commands.getString("HAND"))) {
            return object : GameAction {
                override fun apply(game: Game, current: Player) {
                    val index = Integer.parseInt(string.substring(commands.getString("HAND").length))
                    val cardData = CardManager[current.hand[index]]
                    if (cardData.canCast(current, game)) {
                        current.hand.remove(index)
                        cardData.onCast(current, game)
                    }
                }
            }
        }
        throw RuntimeException("Could not find action \"" + string + "\"")
    }
}
